internal/tasks/convert.go

```diff
@@ -45,67 +46,38 @@ const (
 	ConversionRetryInterval = 2 * time.Minute
 )
 
-// PollForPendingConversions is a JobPoller that finds the next pending
-// conversion job. The returned Job tries to execute the conversion.
-func (c *Context) PollForPendingConversions() (j Job, returnedError error) {
-	defer func() {
-		if returnedError != nil && returnedError != sql.ErrNoRows {
-			labels := prometheus.Labels{"source_payload_type": "early-db-access", "target_payload_type": "early-db-access"}
-			eventConversionFailedCounter.With(labels).Inc()
-		}
-	}()
-
-	//we need a DB transaction for the row-level locking to work correctly
-	tx, err := c.DB.Begin()
-	if err != nil {
-		return nil, err
-	}
-	defer func() {
-		if returnedError != nil {
-			sqlext.RollbackUnlessCommitted(tx)
-		}
-	}()
-
-	//select the next PendingDelivery without converted payload
-	var pd tenso.PendingDelivery
-	err = tx.SelectOne(&pd, selectNextConversionQuery, c.timeNow())
-	if err != nil {
-		if err == sql.ErrNoRows {
-			logg.Debug("no payloads to convert - slowing down...")
-			//nolint:errcheck
-			tx.Rollback() //avoid the log line generated by sqlext.RollbackUnlessCommitted()
-			return nil, sql.ErrNoRows
-		}
-		return nil, err
-	}
-
-	return payloadConvertJob{c, tx, pd}, nil
+func (c *Context) ConversionJob(registerer prometheus.Registerer) jobloop.Job {
+	return (&jobloop.TxGuardedJob[*gorp.Transaction, tenso.PendingDelivery]{
+		Metadata: jobloop.JobMetadata{
+			ReadableName:    "Event conversion",
+			ConcurrencySafe: true, //because "FOR UPDATE SKIP LOCKED" is used
+			CounterOpts: prometheus.CounterOpts{
+				Name: "tenso_event_conversions",
+				Help: "Counter for conversions of event payloads.",
+			},
+			CounterLabels: []string{"source_payload_type", "target_payload_type"},
+		},
+		BeginTx: c.DB.Begin,
+		DiscoverRow: func(_ context.Context, tx *gorp.Transaction, _ prometheus.Labels) (pd tenso.PendingDelivery, err error) {
+			err = tx.SelectOne(&pd, selectNextConversionQuery, c.timeNow())
+			return pd, err
+		},
+		ProcessRow: c.processConversion,
+	}).Setup(registerer)
 }
 
-type payloadConvertJob struct {
-	c  *Context
-	tx *gorp.Transaction
-	pd tenso.PendingDelivery
-}
-
-// Execute implements the Job interface.
-func (j payloadConvertJob) Execute() (returnedError error) {
-	c, tx, pd := j.c, j.tx, j.pd
-
-	defer sqlext.RollbackUnlessCommitted(tx)
-
+func (c *Context) processConversion(_ context.Context, tx *gorp.Transaction, pd tenso.PendingDelivery, labels prometheus.Labels) (returnedError error) {
 	var event tenso.Event
 
 	defer func() {
-		labels := prometheus.Labels{"source_payload_type": event.PayloadType, "target_payload_type": pd.PayloadType}
+		labels["source_payload_type"] = event.PayloadType
+		labels["target_payload_type"] = pd.PayloadType
 		if event.PayloadType == "" { //because we did not get to loading it
 			labels["source_payload_type"] = "early-db-access"
 		}
```

Comment:
This is now prefilled by the library. As a general remark that also applies to the other job, it is now recommended to fill labels within the regular flow of the ProcessRow function, as soon as the label values become available. (If processing fails before that point, the `early-db-access` default value is retained.)

Reply:
Can/Should we assert that somehow by the library?

Reply:
Also shuffled the label filling to the earliest safe position.